lib/src/command/add.dart

```diff
@@ -0,0 +1,326 @@
+  String get gitPath => argResults['git-path'];
+  String get gitRef => argResults['git-ref'];
+  String get hostUrl => argResults['hosted-url'];
+  String get path => argResults['path'];
+  String get sdk => argResults['sdk'];
+
+  bool get hasGitOptions => gitUrl != null || gitRef != null || gitPath != null;
+  bool get hasHostOptions => hostUrl != null;
+
+  AddCommand() {
+    argParser.addFlag('dev',
+        abbr: 'd',
+        negatable: false,
+        help: 'Adds package to the development dependencies instead.');
+
+    argParser.addOption('git-url', help: 'Git URL of the package');
+    argParser.addOption('git-ref',
+        help: 'Git branch or commit to be retrieved');
+    argParser.addOption('git-path', help: 'Path of git package in repository');
+    argParser.addOption('hosted-url', help: 'URL of package host server');
+    argParser.addOption('path', help: 'Local path');
+    argParser.addOption('sdk', help: 'SDK source for package');
+
+    argParser.addFlag('offline',
+        help: 'Use cached packages instead of accessing the network.');
+
+    argParser.addFlag('dry-run',
+        abbr: 'n',
+        negatable: false,
+        help: "Report what dependencies would change but don't change any.");
+
+    argParser.addFlag('precompile',
+        help: 'Precompile executables in immediate dependencies.');
+  }
+
+  @override
+  Future run() async {
+    if (argResults.rest.isEmpty) {
+      usageException('Must specify a package to be added.');
+    }
+
+    final packageInformation = _parsePackage(argResults.rest.first);
+    final package = packageInformation.first;
+
+    /// Perform version resolution in-memory.
+    final updatedPubSpec =
+        await _addPackageToPubspec(entrypoint.root.pubspec, package);
+    final solveResult = await resolveVersions(
+        SolveType.GET, cache, Package.inMemory(updatedPubSpec));
+
+    final resultPackage = solveResult.packages
+        .firstWhere((packageId) => packageId.name == package.name);
+
+    /// Assert that [resultPackage] is within the original user's expectations.
+    if (package.constraint != null &&
+        !package.constraint.allows(resultPackage.version)) {
+      dataError('${package.name} resolved to ${resultPackage.version} which '
+          'does not match the input ${package.constraint}! Exiting.');
+    }
+
+    /// Update the pubspec. It is necessary to do this before calling
+    /// [acquireDependencies] (as opposed to calling [acquireDependencies] on
+    /// the in-memory pubspec and writing it later) because there will be
+    /// conflicts in other parts of [pub] if the pubspec is newer than
+    /// `pubspec.lock`.
+    _updatePubspec(resultPackage, packageInformation);
+
+    /// Create a new [Entrypoint] since we have to reprocess the updated
+    /// pubspec file.
+    await Entrypoint.current(cache).acquireDependencies(SolveType.GET,
+        dryRun: argResults['dry-run'], precompile: argResults['precompile']);
```

Comment:
Do we need to do acquireDependencies if `--dry-run` was passed?

Reply:
hm, I was thinking that this would be in line with what `pub get` does with the `dry-run` flag at the moment, showing what packages will be changed if this new package will be added.

I did mess up the implementation here though - I should probably be having two paths, with the dry run one not updating the actual pubspec and instead calling `acquireDependencies` from a new entrypoint. 

Let me know what your thoughts are on dry-run and I'll make the changes accordingly!

Reply:
I first thought it would be enough to show the changes made to your pubspec, but you are right - other packages might change their locked version.